src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -645,14 +649,12 @@ public StringTokenizer Split(char[] chars)
         /// <returns></returns>
         public static bool IsNullOrEmpty(StringSegment value)
         {
-            bool res = false;
-
             if (!value.HasValue || value.Length == 0)
             {
-                res = true;
+                return true;
             }
 
-            return res;
+            return false;
         }
```

Comment:
It looks like the old version of this method was deliberately written the way it was. It opts into performing two conditional jumps instead of just one + `sete al; movzx rax, al`, as would be the case for this when written naturally: `IsNullOrEmptyNatural(StringSegment value) => !value.HasValue || value.Length == 0;`.

The new version here ends up 1 byte longer due to a second `ret`:
```asm
; Old
       xor      eax, eax
       cmp      gword ptr [rcx], 0
       je       SHORT G_M38619_IG04
						;; bbWeight=1    PerfScore 3.25
G_M38619_IG03:
       cmp      dword ptr [rcx+12], 0
       jne      SHORT G_M38619_IG05
						;; bbWeight=0.50 PerfScore 1.50
G_M38619_IG04:
       mov      eax, 1
						;; bbWeight=0.50 PerfScore 0.12
G_M38619_IG05:
       ret
; Total bytes of code 20, prolog size 0, PerfScore 7.88, instruction count 7
 

; New
       cmp      gword ptr [rcx], 0
       je       SHORT G_M16768_IG04
						;; bbWeight=1    PerfScore 3.00
G_M16768_IG03:
       cmp      dword ptr [rcx+12], 0
       jne      SHORT G_M16768_IG06
						;; bbWeight=0.50 PerfScore 1.50
G_M16768_IG04:
       mov      eax, 1
						;; bbWeight=0.50 PerfScore 0.12
G_M16768_IG05:
       ret      
						;; bbWeight=0.50 PerfScore 0.50
G_M16768_IG06:
       xor      eax, eax
						;; bbWeight=0.50 PerfScore 0.12
G_M16768_IG07:
       ret
; Total bytes of code 21, prolog size 0, PerfScore 7.85, instruction count 8
```
For ARM64, the regression seems to be from `48 bytes` to `56 bytes` (I cannot comment more on that because I do not know ARM's ASM well enough).

That said, when writing this little method (to check the effects on inlining):
```C#
public static int GetLength(StringSegment segment)
{
    if (IsNullOrEmptyNew(segment))
    {
        return 0;
    }

    return segment.Length;
}
```
I saw no differences in codegen between the the new and old versions (for `x64`). Interestingly enough, using the "natural" version of `IsNullOrEmpty` resulted in a significant regression: `23 bytes` -> `31 bytes `, which might just explain the reason behind writing this method the way is has been written.

I severely doubt this one extra `ret` will result in any kind of measurable regressions, especially as it seems it can be eliminated by inlining, so this little investigation of mine is probably of little relevance. Let's see what people with more knowledge about these things than me will say.

Reply:
FYI https://github.com/dotnet/extensions/pull/324
Since then the JIT improved, so some observations from 2018 may be different now. But the code was written with micro-optimizations in mind (although now I don't remember every rationale in details anymore).

Reply:
I can revert IsNullOrEmpty changes :D
By the way, @SingleAccretion, how can I get same asm statistic with bytes count and perfscores?

Reply:
@hrrrrustic https://github.com/egorbo/disasmo is the tool I use. It is incredibly useful in validating these little things.

Reply:
@SingleAccretion ugh I tried it, but it didn't react on my local runtime repo path. It just endlessly asked me to set the path ;(

Reply:
Eh, that's quite unfortuanate. What does the path look like? `C:\Users\Accretion\source\github\runtime` seems to work for me.

Reply:
```D:\Development\VisualStudio\OpenSource\runtime```
Tried to copy into ```C:\```, but it still didn't worl :(

Reply:
@EgorBo Какие подводные могут быть? Просто не реагирует на введенный путь. 5 дотнет стоит, пробовал и на библиотеке, и на консоле

Reply:
Just a sanity check, you do have the `clr -c Checked` subset built, right?

Reply:
I am not seeing this is worth it to change. I suggest just revert this part of the change.

---
In reply to: [553458284](https://github.com/dotnet/runtime/pull/46432#discussion_r553458284) [](ancestors = 553458284)